### PR TITLE
feat(review): add goal progress chart to buzz review (#170)

### LINKED ISSUE #170: Display goal graphs in `buzz review`
## Overview
Display goal graphs in `buzz review` to provide visual feedback on goal progress, showing both datapoints and the yellow brick road.

### ASCII Chart Visualization

Add an in-terminal ASCII chart that displays goal progress with the following features:

**Required Goal Data Fields** (add to `Goal` struct in `beeminder.go`):
- `Yaw` (int): Indicates which side is good (+1 = above road, -1 = below road)
- `Dir` (int): Direction of road slope
- `Kyoom` (bool): Whether goal is cumulative/auto-summing
- `Roadall` ([][]any): Yellow brick road segments [date, value, rate]
- `Tmin` (string): Min date for graph view (yyyy-mm-dd format)
- `Tmax` (string): Max date for graph view (yyyy-mm-dd format)

**Chart Features**:
- ✅ Shows datapoints within the `tmin` to `tmax` timeframe
- ✅ Fallback to last 30 days if tmin/tmax not available
- ✅ Displays yellow brick road target line
- ✅ Handles cumulative vs non-cumulative goals correctly
  - Cumulative: sums datapoint values progressively
  - Non-cumulative: uses actual datapoint values
- ✅ Visual indicators:
  - `●` = datapoint on good side of road
  - `○` = datapoint on bad side of road
  - `─` = yellow brick road target line
  - `█` = datapoint exactly on target
- ✅ Shows goal type and side requirement in header
- ✅ Displays timeframe information in header
- ✅ Includes legend for clarity

**Implementation Location**:
- Add `renderGoalChart()` function to handle chart rendering logic
- Add `getRoadValuesForTimeframe()` to calculate road values for datapoint timestamps
- Add `getRoadValueAtTime()` for road value interpolation
- Integrate into `review.go` View() method after goal details section

## Technical Details

**Data Flow**:
1. Fetch goals with extended fields (yaw, kyoom, roadall, tmin, tmax)
2. Parse tmin/tmax date strings (yyyy-mm-dd format) to determine timeframe
3. Filter datapoints to only those within the timeframe
4. Process datapoints based on cumulative setting
5. Calculate road values for visualization timeframe
6. Render ASCII chart with proper scaling
7. Display in review interface with styled border

**Edge Cases to Handle**:
- Goals with no datapoints
- Goals with no datapoints in the specified timeframe
- Goals with no road data
- Missing or invalid tmin/tmax values (fallback to last 30 days)
- Cumulative goals requiring value summation
- Different road segment formats (timestamps, daystamps)
- Terminal width constraints

## Success Criteria

- [ ] Chart displays correctly for both cumulative and non-cumulative goals
- [ ] Visual distinction between datapoints on good/bad side of road
- [ ] Yellow brick road target line is accurately plotted
- [ ] Chart respects tmin/tmax timeframe from Beeminder API
- [ ] Fallback to last 30 days when tmin/tmax unavailable
- [ ] Chart scales properly to terminal width
- [ ] Works across different terminal sizes

## Related Issues

Additional enhancements have been extracted to separate issues for better tracking.

### PR DESCRIPTION
**Pull request by narthur**
Fresh reimplementation of the goal-progress chart feature (replacing the stale #173, which had drifted ~50 files behind `main` and conflicted with the recent rate-display work).

## Screenshots

`buzz review` with the progress chart — datapoints (blue) over the bright red line (red), with a date-labeled x-axis.

| Cumulative (Do More) | Non-cumulative (Do More) |
|---|---|
| buzz review chart — cumulative goal | buzz review chart — non-cumulative goal |

## What it does
`buzz review` now draws an ASCII chart of each goal's recent progress beneath the existing details: the goal's **datapoints (blue)** against its **bright red line (red)**.

```
  Goal: active
  Pledge:      $5.00
  Rate:        60 minutes / day
  Goal Progress Chart - Do More (Cumulative)
  Timeframe: May 7 to Jun 6, 2026
 63989 ┤                                          ╭─────────╭─────
   ...
 62354 ┼╯
              Blue: datapoints, Red: bright red line
```

- **Timeframe**: the goal's graph window (`tmin`..`tmax`) when an end date is set, else the last 30 days. (`tmax` is null on most goals, so the 30-day window is the common path.)
- **Cumulative (`kyoom`) goals** plot the running total, with a synthetic anchor at the window start so the line begins where the goal actually stands rather than at zero.
- **Road interpolation** walks `roadall`'s piecewise schedule, materialising rate-only rows, extrapolating before the anchor, and bailing safely on ambiguous rows or untranslatable runits.

## Design notes
- Chart logic is self-contained in **`chart.go`** with a full unit suite in **`chart_test.go`** (17 tests: cumulative/non-cumulative, do-more/do-less, fallback window, tmax-day boundary, road edge cases).
- Renders **synchronously** in `review.View()` — no async loading machinery. The data is already loaded by the existing `FetchGoalsWithDatapoints`, which now simply **retains** the `roadall`/`tmin`/`tmax`/`kyoom` fields it was previously discarding. Integration footprint is a single `if` in `View()` plus the field-retention in the fetch.
- New dependency: `github.com/guptarohit/asciigraph`.

## Verification
- `go build`, full `go test ./...`, `go vet`, `gofmt` all clean.
- Ran `buzz review` against live Beeminder data: chart renders correctly for cumulative do-more goals; confirmed `roadall` present on all 63 goals and datapoints on 62.

## Known limitation (pre-existing, not introduced here)
`buzz review` already fetches every goal's datapoints up front (`FetchGoalsWithDatapoints`), which for this account (~63 goals) takes ~50s before the TUI appears. This predates the chart feature — the detail view depends on the same fetch. Worth a follow-up to lazy-load per goal, but out of scope here.

Closes #170

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Visual ASCII progress charts in goal details, with timeframe bounds and cumulative/reduction handling for accurate rendering.
* **Bug Fixes**
  * Charts now appear consistently for goals fetched via bulk listing.
* **Tests**
  * Added comprehensive tests covering chart rendering, x‑axis alignment, and road/progression logic.
* **Chores**
  * Added a plotting/graphing library dependency.

### CODE REVIEW

File: chart.go [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+
+	start, err := time.ParseInLocation("2006-01-02", goal.Tmin, time.Local)
+	if err != nil {
+		start = now.AddDate(0, 0, -30)
+	}
+	end, err = time.ParseInLocation("2006-01-02", goal.Tmax, time.Local)
+	if err != nil {
+		return start, now
+	}
+	end = end.Add(24*time.Hour - time.Second)
```
[the remark this answers is not in the document]
Comment (narthur) [reply chart.go]: Good catch — fixed. Now computes end-of-day as the start of the next calendar day minus one second via `time.Date(..., Day()+1, ...)`, which is DST-aware, instead of adding a fixed 24h. Added a regression test (`TestChartTimeframeTmaxAcrossDSTFallBack`) covering a 25h fall-back day in America/New_York.

File: chart.go [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+		case !dpTime.After(endTime):
+			sum += dp.Value
+			processed = append(processed, timedValue{timestamp: dp.Timestamp, value: sum})
+		}
+		// Datapoints after endTime are ignored.
+	}
+
+	if len(processed) == 0 && startSum == 0 {
+		return nil
+	}
```
[the remark this answers is not in the document]
Comment (narthur) [reply chart.go]: Fixed — `processCumulative` now returns nil whenever no datapoints fall inside the window, regardless of the carry-over total, so no flat dataless line is drawn. Matches the docstring/caller contract. Added `TestRenderGoalChartCumulativeAllBeforeWindow`.

File: chart.go [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+
+	graphOutput := asciigraph.PlotMany([][]float64{datapointValues, roadValues},
+		asciigraph.Height(chartHeight),
+		asciigraph.Width(chartWidth),
+		asciigraph.SeriesColors(asciigraph.Blue, asciigraph.Red),
+		asciigraph.Caption("Blue: datapoints, Red: bright red line"),
+	)
+
+	chart.WriteString(graphOutput)
+	chart.WriteString("\n")
```
[the remark this answers is not in the document]
Comment (narthur) [reply chart.go]: Fixed — the plot and date axis are now indented 2 spaces (via a new `indentLines` helper) to match the header/caption/details padding, so the chart no longer renders left-shifted. The y-axis gutter is measured before indenting so axis alignment is preserved.

File: chart_test.go [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+	}
+	if !strings.Contains(chart, "Goal Progress Chart") {
+		t.Error("Expected chart to contain 'Goal Progress Chart'")
+	}
+	if !strings.Contains(chart, "Do More") {
+		t.Error("Expected chart to contain 'Do More'")
+	}
+	if !strings.Contains(chart, "datapoints") && !strings.Contains(chart, "bright red line") {
+		t.Error("Expected chart to contain caption")
+	}
```
[the remark this answers is not in the document]
Comment (narthur) [reply chart_test.go]: Good catch — changed to `||` so the assertion fails if either caption substring is missing.

File: chart_test.go [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+
+	chart := renderGoalChart(goal, 80)
+	if chart == "" {
+		t.Fatal("Expected non-empty chart for cumulative goal")
+	}
+	if !strings.Contains(chart, "Cumulative") {
+		t.Error("Expected chart to indicate cumulative goal")
+	}
+}
```
[the remark this answers is not in the document]
Comment (narthur) [reply chart_test.go:86]: Added `TestProcessCumulativeValues` — directly asserts the cumulative values (anchor + running totals), including the carry-over case where a pre-window datapoint feeds the anchor but isn't itself plotted. A summation bug would now be caught.

File: chart.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (narthur) [reply chart.go]: Resolving without a change. Beeminder datapoint timestamps are whole Unix seconds (`Datapoint.Timestamp int64`, always read via `time.Unix(ts, 0)`), so a sub-second time like 23:59:59.500 can't occur — the finest granularity is 23:59:59.000, which the current bound includes (`!dpTime.After(end)`). The earlier fix also replaced the original `24h - 1s` with a DST-correct `time.Date(..., Day()+1, ...).Add(-time.Second)`, so the substantive day-boundary concern is already handled. Switching to an exclusive next-day bound would be equivalent here but isn't needed given integer-second timestamps.

File: go.mod [hunk 6]
```diff
@@ -7,6 +7,7 @@ require (
 	github.com/charmbracelet/bubbletea v1.3.10
 	github.com/charmbracelet/lipgloss v1.1.0
 	github.com/charmbracelet/x/term v0.2.1
+	github.com/guptarohit/asciigraph v0.7.3
```
[the remark this answers is not in the document]
Comment (narthur) [reply go.mod]: Bumped to v0.9.0. Build + full test suite pass and the rendered chart is unchanged in practice (verified an offline render: same axis runes, series colors, and layout).
